File: Python3/valid_palindrome.py
import string

class Solution:
    def isPalindrome(self, s: str) -> bool:
        if not s: return True
        
        allow = set(string.ascii_lowercase + string.digits)
        s = [c for c in s.lower() if c in allow]
        
        i, j = 0, len(s)-1        
        while i < j:
            if s[i] != s[j]: return False
            i += 1
            j -= 1
            
        return True

File: C++/valid_palindrome.cpp
class Solution {
public:
    string toLowerCase(string str) {
        for(char & c : str) if (c >= 'A' && c <= 'Z') c += 32;        
        return (str);
    }
    
    bool isPalindrome(string s) {
        s = toLowerCase(s);
        
        int i = 0, j = s.size()-1;
        while(i < j) {
            while(i < j and not isalnum(s[i])) ++i;
            while(i < j and not isalnum(s[j])) --j;
            if(i < j and s[i] != s[j]) return false;
            i++, j--;
        }
        return true;
    }
};
